File: macos/MISILNative/Sources/MISILNative/RelayService.swift
import CryptoKit
import Foundation
import Security

struct NativeRelayIdentity: Codable, Equatable, Sendable {
    let version: Int
    let baseURL: String
    let roomID: String
    let accessToken: String
    let encryptionKey: String
    let deviceID: String
    let displayName: String
    let createdAt: String

    var accessCode: String {
        let secret = SharedRoomSecret(v: 1, r: roomID, t: accessToken, k: encryptionKey)
        guard let data = try? JSONEncoder().encode(secret) else { return "" }
        return data.base64URLString
    }
}

struct NativeRelayMessage: Identifiable, Equatable, Sendable {
    let id: UUID
    let content: String
    let createdAt: Date
    let senderID: UUID
    let senderName: String
}

enum RelayServiceError: LocalizedError {
    case invalidEndpoint
    case invalidAccessCode
    case invalidResponse
    case relay(String)

    var errorDescription: String? {
        switch self {
        case .invalidEndpoint: "La dirección del relay no es válida."
        case .invalidAccessCode: "El código privado no es válido."
        case .invalidResponse: "El relay devolvió una respuesta inesperada."
        case .relay(let message): message
        }
    }
}

actor RelayService {
    private let session: URLSession

    init(session: URLSession = .shared) {
        self.session = session
    }

    func createRoom(displayName: String, baseURL: String) async throws -> NativeRelayIdentity {
        let identity = NativeRelayIdentity(
            version: 1,
            baseURL: try normalizedBaseURL(baseURL),
            roomID: UUID().uuidString.lowercased(),
            accessToken: try randomSecret(),
            encryptionKey: try randomSecret(),
            deviceID: UUID().uuidString.lowercased(),
            displayName: displayName.trimmingCharacters(in: .whitespacesAndNewlines),
            createdAt: ISO8601DateFormatter().string(from: Date())
        )
        let body = RoomRegistration(
            roomId: identity.roomID,
            tokenHash: SHA256.hash(data: Data(identity.accessToken.utf8)).hexString
        )
        _ = try await request(identity: identity, path: "/api/relay/rooms", method: "POST", body: body, authorized: false) as EmptyResponse
        return identity
    }

    func joinRoom(code: String, displayName: String, baseURL: String) async throws -> NativeRelayIdentity {
        guard let secretData = Data(base64URL: code.trimmingCharacters(in: .whitespacesAndNewlines)),
              let secret = try? JSONDecoder().decode(SharedRoomSecret.self, from: secretData),
              secret.v == 1,
              Data(base64URL: secret.k)?.count == 32
        else { throw RelayServiceError.invalidAccessCode }

        let identity = NativeRelayIdentity(
            version: 1,
            baseURL: try normalizedBaseURL(baseURL),
            roomID: secret.r,
            accessToken: secret.t,
            encryptionKey: secret.k,
            deviceID: UUID().uuidString.lowercased(),
            displayName: displayName.trimmingCharacters(in: .whitespacesAndNewlines),
            createdAt: ISO8601DateFormatter().string(from: Date())
        )
        _ = try await messages(identity: identity)
        return identity
    }

    func send(_ content: String, identity: NativeRelayIdentity) async throws -> NativeRelayMessage {
        let messageID = UUID()
        let createdAt = ISO8601DateFormatter().string(from: Date())
        let senderID = UUID(uuidString: identity.deviceID) ?? UUID()
        let payload = RelayPayload(
            id: messageID.uuidString.lowercased(),
            content: content,
            createdAt: createdAt,
            sender: .init(id: senderID.uuidString.lowercased(), displayName: identity.displayName)
        )
        let plain = try JSONEncoder().encode(payload)
        guard let keyData = Data(base64URL: identity.encryptionKey) else { throw RelayServiceError.invalidAccessCode }
        let sealed = try AES.GCM.seal(plain, using: SymmetricKey(data: keyData))
        let encrypted = sealed.ciphertext + sealed.tag
        let envelope = OutgoingEnvelope(
            roomId: identity.roomID,
            id: payload.id,
            ciphertext: encrypted.base64URLString,
            iv: Data(sealed.nonce).base64URLString,
            createdAt: createdAt
        )
        _ = try await request(identity: identity, path: "/api/relay/messages", method: "POST", body: envelope, authorized: true) as EmptyResponse
        return NativeRelayMessage(id: messageID, content: content, createdAt: Date(), senderID: senderID, senderName: identity.displayName)
    }

    func messages(identity: NativeRelayIdentity) async throws -> [NativeRelayMessage] {
        guard let room = identity.roomID.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) else {
            throw RelayServiceError.invalidAccessCode
        }
        let response: EnvelopeList = try await request(
            identity: identity,
            path: "/api/relay/messages?roomId=\(room)",
            method: "GET",
            body: Optional<EmptyResponse>.none,
            authorized: true
        )
        guard let keyData = Data(base64URL: identity.encryptionKey) else { throw RelayServiceError.invalidAccessCode }
        let key = SymmetricKey(data: keyData)
        let formatter = ISO8601DateFormatter()

        return response.messages.compactMap { envelope in
            guard let encrypted = Data(base64URL: envelope.ciphertext), encrypted.count > 16,
                  let nonceData = Data(base64URL: envelope.iv),
                  let nonce = try? AES.GCM.Nonce(data: nonceData)
            else { return nil }
            let ciphertext = encrypted.dropLast(16)
            let tag = encrypted.suffix(16)
            guard let box = try? AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag),
                  let plain = try? AES.GCM.open(box, using: key),
                  let payload = try? JSONDecoder().decode(RelayPayload.self, from: plain),
                  payload.id == envelope.id,
                  let id = UUID(uuidString: payload.id),
                  let senderID = UUID(uuidString: payload.sender.id)
            else { return nil }
            return NativeRelayMessage(
                id: id,
                content: payload.content,
                createdAt: formatter.date(from: payload.createdAt) ?? Date(),
                senderID: senderID,
                senderName: payload.sender.displayName
            )
        }.sorted { $0.createdAt < $1.createdAt }
    }

    private func request<Response: Decodable, Body: Encodable>(
        identity: NativeRelayIdentity,
        path: String,
        method: String,
        body: Body?,
        authorized: Bool
    ) async throws -> Response {
        guard let url = URL(string: identity.baseURL + path) else { throw RelayServiceError.invalidEndpoint }
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.timeoutInterval = 20
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let body {
            request.httpBody = try JSONEncoder().encode(body)
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }
        if authorized { request.setValue("Bearer \(identity.accessToken)", forHTTPHeaderField: "Authorization") }
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw RelayServiceError.invalidResponse }
        if !(200..<300).contains(http.statusCode) {
            let failure = try? JSONDecoder().decode(RelayFailure.self, from: data)
            throw RelayServiceError.relay(failure?.error ?? "El relay respondió con \(http.statusCode).")
        }
        if Response.self == EmptyResponse.self, data.isEmpty == false,
           let decoded = try? JSONDecoder().decode(Response.self, from: data) { return decoded }
        if Response.self == EmptyResponse.self { return EmptyResponse() as! Response }
        return try JSONDecoder().decode(Response.self, from: data)
    }

    private func normalizedBaseURL(_ value: String) throws -> String {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines).trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        guard let url = URL(string: trimmed), ["http", "https"].contains(url.scheme), url.host != nil else {
            throw RelayServiceError.invalidEndpoint
        }
        return trimmed
    }

    private func randomSecret() throws -> String {
        var bytes = [UInt8](repeating: 0, count: 32)
        let status = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        guard status == errSecSuccess else { throw StorageSetupError.keychainFailure(status) }
        return Data(bytes).base64URLString
    }
}

private struct SharedRoomSecret: Codable { let v: Int; let r: String; let t: String; let k: String }
private struct RoomRegistration: Codable { let roomId: String; let tokenHash: String }
private struct OutgoingEnvelope: Codable { let roomId: String; let id: String; let ciphertext: String; let iv: String; let createdAt: String }
private struct EnvelopeList: Codable { let messages: [IncomingEnvelope] }
private struct IncomingEnvelope: Codable { let id: String; let ciphertext: String; let iv: String; let createdAt: String }
private struct RelayPayload: Codable {
    let id: String
    let content: String
    let createdAt: String
    let sender: Sender
    struct Sender: Codable { let id: String; let displayName: String }
}
private struct RelayFailure: Codable { let error: String }
private struct EmptyResponse: Codable { init() {} }

private extension Data {
    var base64URLString: String {
        base64EncodedString().replacingOccurrences(of: "+", with: "-").replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "")
    }

    init?(base64URL: String) {
        var normalized = base64URL.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        normalized += String(repeating: "=", count: (4 - normalized.count % 4) % 4)
        self.init(base64Encoded: normalized)
    }
}

private extension SHA256.Digest {
    var hexString: String { map { String(format: "%02x", $0) }.joined() }
}

File: windows/MISILNative/Services/RelayService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using MISILNative.Models;

namespace MISILNative.Services
{
    public class RelayService
    {
        private static readonly HttpClient HttpClient = new() { Timeout = TimeSpan.FromSeconds(20) };

        public async Task<NativeRelayIdentity> CreateRoomAsync(string displayName, string baseUrl)
        {
            string normalizedUrl = NormalizeBaseUrl(baseUrl);
            var identity = new NativeRelayIdentity
            {
                Version = 1,
                BaseUrl = normalizedUrl,
                RoomId = Guid.NewGuid().ToString().ToLowerInvariant(),
                AccessToken = RandomSecretBase64Url(32),
                EncryptionKey = RandomSecretBase64Url(32),
                DeviceId = Guid.NewGuid().ToString().ToLowerInvariant(),
                DisplayName = displayName.Trim(),
                CreatedAt = DateTime.UtcNow.ToString("o")
            };

            var body = new RoomRegistration
            {
                RoomId = identity.RoomId,
                TokenHash = ComputeSha256Hex(identity.AccessToken)
            };

            await SendRequestAsync(identity, "/api/relay/rooms", HttpMethod.Post, body, authorized: false);
            return identity;
        }

        public async Task<NativeRelayIdentity> JoinRoomAsync(string code, string displayName, string baseUrl)
        {
            var secretBytes = Base64UrlDecode(code.Trim());
            var secretJson = Encoding.UTF8.GetString(secretBytes);
            var secret = JsonSerializer.Deserialize<SharedRoomSecret>(secretJson);

            if (secret == null || secret.V != 1 || Base64UrlDecode(secret.K).Length != 32)
            {
                throw new InvalidOperationException("El código privado no es válido o está incompleto.");
            }

            var identity = new NativeRelayIdentity
            {
                Version = 1,
                BaseUrl = NormalizeBaseUrl(baseUrl),
                RoomId = secret.R,
                AccessToken = secret.T,
                EncryptionKey = secret.K,
                DeviceId = Guid.NewGuid().ToString().ToLowerInvariant(),
                DisplayName = displayName.Trim(),
                CreatedAt = DateTime.UtcNow.ToString("o")
            };

            // Test fetching messages to verify room and credentials
            await FetchMessagesAsync(identity);
            return identity;
        }

        public async Task<NativeMessage> SendAsync(string content, NativeRelayIdentity identity)
        {
            var messageId = Guid.NewGuid();
            var createdAt = DateTime.UtcNow.ToString("o");
            Guid senderId = Guid.TryParse(identity.DeviceId, out var sid) ? sid : Guid.NewGuid();

            var payload = new RelayPayload
            {
                Id = messageId.ToString().ToLowerInvariant(),
                Content = content,
                CreatedAt = createdAt,
                Sender = new RelaySender
                {
                    Id = senderId.ToString().ToLowerInvariant(),
                    DisplayName = identity.DisplayName
                }
            };

            byte[] plainBytes = JsonSerializer.SerializeToUtf8Bytes(payload);
            byte[] keyBytes = Base64UrlDecode(identity.EncryptionKey);

            byte[] nonce = new byte[12]; // 96-bit standard nonce for AES-GCM
            RandomNumberGenerator.Fill(nonce);

            byte[] ciphertext = new byte[plainBytes.Length];
            byte[] tag = new byte[16]; // 128-bit authentication tag

            using (var aesGcm = new AesGcm(keyBytes, 16))
            {
                aesGcm.Encrypt(nonce, plainBytes, ciphertext, tag);
            }

            byte[] combinedCiphertext = new byte[ciphertext.Length + tag.Length];
            Buffer.BlockCopy(ciphertext, 0, combinedCiphertext, 0, ciphertext.Length);
            Buffer.BlockCopy(tag, 0, combinedCiphertext, ciphertext.Length, tag.Length);

            var envelope = new OutgoingEnvelope
            {
                RoomId = identity.RoomId,
                Id = payload.Id,
                Ciphertext = Base64UrlEncode(combinedCiphertext),
                Iv = Base64UrlEncode(nonce),
                CreatedAt = createdAt
            };

            await SendRequestAsync(identity, "/api/relay/messages", HttpMethod.Post, envelope, authorized: true);

            return new NativeMessage
            {
                Id = messageId,
                Content = content,
                CreatedAt = DateTime.UtcNow,
                SenderId = senderId,
                SenderName = identity.DisplayName
            };
        }

        public async Task<List<NativeMessage>> FetchMessagesAsync(NativeRelayIdentity identity)
        {
            string path = $"/api/relay/messages?roomId={Uri.EscapeDataString(identity.RoomId)}";
            string responseJson = await SendRequestAsync(identity, path, HttpMethod.Get, (object?)null, authorized: true);

            var envelopeList = JsonSerializer.Deserialize<EnvelopeList>(responseJson);
            if (envelopeList?.Messages == null) return new List<NativeMessage>();

            byte[] keyBytes = Base64UrlDecode(identity.EncryptionKey);
            var result = new List<NativeMessage>();

            foreach (var envelope in envelopeList.Messages)
            {
                try
                {
                    byte[] combinedCipher = Base64UrlDecode(envelope.Ciphertext);
                    if (combinedCipher.Length <= 16) continue;

                    byte[] nonce = Base64UrlDecode(envelope.Iv);
                    if (nonce.Length != 12) continue;

                    int cipherLen = combinedCipher.Length - 16;
                    byte[] ciphertext = new byte[cipherLen];
                    byte[] tag = new byte[16];
                    Buffer.BlockCopy(combinedCipher, 0, ciphertext, 0, cipherLen);
                    Buffer.BlockCopy(combinedCipher, cipherLen, tag, 0, 16);

                    byte[] decryptedPlain = new byte[cipherLen];
                    using (var aesGcm = new AesGcm(keyBytes, 16))
                    {
                        aesGcm.Decrypt(nonce, ciphertext, tag, decryptedPlain);
                    }

                    var payload = JsonSerializer.Deserialize<RelayPayload>(decryptedPlain);
                    if (payload != null && Guid.TryParse(payload.Id, out var id))
                    {
                        Guid? sId = Guid.TryParse(payload.Sender.Id, out var sid) ? sid : null;
                        DateTime.TryParse(payload.CreatedAt, out var dt);

                        result.Add(new NativeMessage
                        {
                            Id = id,
                            Content = payload.Content,
                            CreatedAt = dt != default ? dt : DateTime.UtcNow,
                            SenderId = sId,
                            SenderName = payload.Sender.DisplayName
                        });
                    }
                }
                catch { }
            }

            return result.OrderBy(m => m.CreatedAt).ToList();
        }

        private async Task<string> SendRequestAsync(
            NativeRelayIdentity identity,
            string path,
            HttpMethod method,
            object? body,
            bool authorized)
        {
            var url = $"{identity.BaseUrl}{path}";
            using var request = new HttpRequestMessage(method, url);
            request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            if (authorized)
            {
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", identity.AccessToken);
            }

            if (body != null)
            {
                var json = JsonSerializer.Serialize(body);
                request.Content = new StringContent(json, Encoding.UTF8, "application/json");
            }

            using var response = await HttpClient.SendAsync(request);
            string responseBody = await response.Content.ReadAsStringAsync();

            if (!response.IsSuccessStatusCode)
            {
                try
                {
                    var failure = JsonSerializer.Deserialize<RelayFailure>(responseBody);
                    if (!string.IsNullOrEmpty(failure?.Error))
                    {
                        throw new InvalidOperationException(failure.Error);
                    }
                }
                catch (JsonException) { }

                throw new InvalidOperationException($"El relay respondió con error HTTP {(int)response.StatusCode}.");
            }

            return responseBody;
        }

        private string NormalizeBaseUrl(string value)
        {
            string trimmed = value.Trim().TrimEnd('/');
            if (!Uri.TryCreate(trimmed, UriKind.Absolute, out var uri) ||
                (uri.Scheme != "http" && uri.Scheme != "https"))
            {
                throw new InvalidOperationException("La dirección URL del relay no es válida.");
            }
            return trimmed;
        }

        private string RandomSecretBase64Url(int size = 32)
        {
            byte[] bytes = new byte[size];
            RandomNumberGenerator.Fill(bytes);
            return Base64UrlEncode(bytes);
        }

        private string ComputeSha256Hex(string input)
        {
            byte[] inputBytes = Encoding.UTF8.GetBytes(input);
            byte[] hash = SHA256.HashData(inputBytes);
            return Convert.ToHexString(hash).ToLowerInvariant();
        }

        private string Base64UrlEncode(byte[] bytes)
        {
            return Convert.ToBase64String(bytes)
                .Replace("+", "-")
                .Replace("/", "_")
                .TrimEnd('=');
        }

        private byte[] Base64UrlDecode(string value)
        {
            string incoming = value.Replace("-", "+").Replace("_", "/");
            switch (incoming.Length % 4)
            {
                case 2: incoming += "=="; break;
                case 3: incoming += "="; break;
            }
            return Convert.FromBase64String(incoming);
        }
    }
}
